Warn that UseStartup/Configure/ConfigureWebHost not supported on WebApplicationBuilder

martincostello opened the pull request:
**PR Title**

Add WebApplicationBuilder analyzers for unsupported methods

**PR Description**

Add analyzers that warn on incorrect usage of `UseStartup()`, `Configure()` and `ConfigureWebHost()` when using `WebApplicationBuilder.Host` and `WebApplicationBuilder.WebHost`.

## TODO

- [x] Review the exact diagnostic IDs to use
- [x] Review the exact diagnostic messages to use
- [x] ~Add tests for existing startup patterns prior to .NET 6 to verify they do not generate false positives~
- [x] ~Refactor analyzer to reduce duplication in detecting similar usage patterns~

Fixes #35814

pranavkm commented on src/Framework/Analyzer/src/WebApplicationBuilder/WebApplicationBuilderAnalyzer.cs:
```diff
@@ -0,0 +1,137 @@
+                return false;
+            }
+
+            return disallowedMethodTypes.Any(type => SymbolEqualityComparer.Default.Equals(type, methodSymbol.ContainingType));
```
Could we use a regular for loop? We should treat all of the analyzers as if they were running in the hot path all the time.

pranavkm commented on src/Framework/Analyzer/src/WebApplicationBuilder/WebApplicationBuilderAnalyzer.cs:
```diff
@@ -0,0 +1,137 @@
+        string disallowedMethodName,
+        params INamedTypeSymbol[] disallowedMethodTypes)
+    {
+        if (!IsDisallowedMethod(methodSymbol, disallowedMethodName, disallowedMethodTypes))
```
This is a great check to start this with.

pranavkm commented on src/Framework/Analyzer/src/WebApplicationBuilder/WebApplicationBuilderAnalyzer.cs:
```diff
@@ -0,0 +1,137 @@
+            return typeInfo.Type as INamedTypeSymbol;
+        }
+
+        return null;
```
Super nit: Could we put this on line 125 (before the local function)?

pranavkm commented on src/Framework/Analyzer/test/WebApplicationBuilder/DisallowConfigureWebHostBuilderConfigureTest.cs:
```diff
@@ -0,0 +1,107 @@
+    public async Task WebApplicationBuilder_WebHostWithConfigureWithContext_ProducesDiagnostics()
+    {
+        // Arrange
+        var source = TestSource.Read(@"
```
I know it doesn't matter to the implementation, but could we add one test with producing diagnostics in a non-top level Program?

martincostello replied:
Sure - didn't think of that.

pranavkm reviewed (commented):
This looks great!

martincostello commented on src/Framework/Analyzer/src/WebApplicationBuilder/DiagnosticDescriptors.cs:
```diff
@@ -0,0 +1,37 @@
+    internal static readonly DiagnosticDescriptor DoNotUseConfigureWebHostWithConfigureHostBuilder = new(
+        "ASP0005",
+        "Do not use ConfigureWebHost with WebApplicationBuilder.Host",
+        "ConfigureWebHost cannot be used with WebApplicationBuilder.Host",
```
Yeah I wasn't sure on the wording that would be wanted, so just went with something basic as a starting point.

davidfowl commented:
I want to feature creep this PR as I realize there are more APIs I want to block but this is awesome

martincostello commented:
> I want to feature creep this PR as I realize there are more APIs

😄 

I'd be happy to add some more similar patterns as a follow-up post-merge if you've got a laundry list.

martincostello commented on src/Framework/Analyzer/src/DiagnosticDescriptors.cs:
```diff
@@ -0,0 +1,46 @@
+        helpLinkUri: "https://aka.ms/aspnet/analyzers");
+
+    internal static readonly DiagnosticDescriptor DoNotUseConfigureWebHostWithConfigureHostBuilder = new(
+        "ASP0005",
```
Where did 4 go... 🤔

Maybe I can't count?

captainsafia replied:
`ASP0004` should map to `DoNotReturnActionResultsFromMapActions`.  I think you might have missed a commit during a rebase or something.

`ASP0005` is also taken in a recent PR that I've merged: https://github.com/dotnet/aspnetcore/blob/d9ac5f4041e7912cfec08656a775b4418f203d76/src/Framework/Analyzer/src/DelegateEndpoints/DiagnosticDescriptors.cs#L29

Unfortunately, there's gonna be a lot of stomping over this source file as we build out analyzers in the space.

martincostello commented on src/Framework/Analyzer/test/WebApplicationBuilder/DisallowConfigureHostBuilderConfigureWebHostTest.cs:
```diff
@@ -0,0 +1,115 @@
+    public static void Main(string[] args)
+    {
+        var builder = WebApplication.CreateBuilder(args);
+        /*MM*/builder.Host.ConfigureWebHost(webHostBuilder => { });
```
Is the marker here supposed to be on the line start instead of at the problematic invocation?

captainsafia replied:
The pattern I've been following is to emit the diagnostic at the point where the problematic invocation exists so if the user clicks on the diagnostic's location in the build output the cursor will be directed to the place where they need to make the fix.

It also makes it easier to write codefixes associated with the analyzer because you can get the syntax at the problematic location right away.

martincostello replied:
Yeah I figured that would be the case. I originally put it before `ConfigureWebHost`, but then the tests failed even though I was just using the syntax given to the analyzer to get the location for the invocation. I'll dig around more and see if I can work out why there's a mismatch when I do that so the warning is in the best place.

martincostello commented on src/Framework/Analyzer/test/WebApplicationBuilder/DisallowConfigureWebHostBuilderConfigureTest.cs:
```diff
@@ -0,0 +1,134 @@
+    public static void Main(string[] args)
+    {
+        var builder = WebApplication.CreateBuilder(args);
+        /*MM*/builder.WebHost.Configure((context, webHostBuilder) => { });
```
Same question.

martincostello commented on src/Framework/Analyzer/test/WebApplicationBuilder/DisallowConfigureWebHostBuilderUseStartupTest.cs:
```diff
@@ -0,0 +1,181 @@
+    public static void Main(string[] args)
+    {
+        var builder = WebApplication.CreateBuilder(args);
+        /*MM*/builder.WebHost.UseStartup<Startup>();
```
Same question.

martincostello commented on src/Framework/Analyzer/src/WebApplicationBuilder/WebApplicationBuilderAnalyzer.cs:
```diff
@@ -0,0 +1,182 @@
+                return;
+            }
+
+            INamedTypeSymbol[] configureTypes = { wellKnownTypes.WebHostBuilderExtensions };
```
Couldn't statically cache as the compilation is needed to get the well-known types, but I guess the lifetime of the class is "long enough" that these locals being reused in the lambda are OK?

davidfowl commented:
@pranavkm you are the gate here

davidfowl commented:
@captainsafia you as well.

martincostello commented on src/Framework/Analyzer/src/WebApplicationBuilder/WebApplicationBuilderAnalyzer.cs:
```diff
@@ -0,0 +1,239 @@
+                    //
+                    // IdentifierNameSyntax finds non-generic methods (e.g. `Foo()`), whereas
+                    // GenericNameSyntax finds generic methods (e.g. `Foo<T>()`).
+                    var methodName = operation.Syntax
+                        .DescendantNodes()
+                        .OfType<SimpleNameSyntax>()
+                        .Where(node => node is IdentifierNameSyntax || node is GenericNameSyntax)
+                        .Where(node => string.Equals(node.Identifier.Value as string, operation.TargetMethod.Name, StringComparison.Ordinal))
+                        .FirstOrDefault();
```
Happy to refactor this if there's a better way I didn't work out for myself.

I assumed that using LINQ here was OK as we're already on the path where a diagnostic is going to be fired.

martincostello commented on src/Framework/Analyzer/test/WebApplicationBuilder/DisallowConfigureHostBuilderConfigureWebHostTest.cs:
```diff
@@ -0,0 +1,240 @@
+using Microsoft.Extensions.Hosting;
+var builder = WebApplication.CreateBuilder();
+builder.Host
+    .ConfigureServices(services => { }) // Because ConfigureServices() returns IHostBuilder, the type gets erased
```
Happy to remove this test if it doesn't add any value. It started as my check the analyzer _did_ work, but then realising that as the receiver type on `ConfigureWebHost()` became `IHostBuilder` it could no longer be detected.

Alternatively I could try and make it actually fire, but that would mean walking up the invocation chain and trying to find the WebApplicationBuilder.Host at the root, but I'm not sure that's worth adding that additional complexity to try and cover so many possible usage patterns.

captainsafia commented on src/Framework/AspNetCoreAnalyzers/src/Analyzers/WebApplicationBuilder/WebApplicationBuilderAnalyzer.cs:
```diff
@@ -0,0 +1,239 @@
+        }
+    }
+
+    // GetReceiverType() adapted from IOperationExtensions.GetReceiverType in dotnet/roslyn-analyzers.
```
Nit: We can move this to `src/Shared/Roslyn`.

martincostello commented on src/Framework/AspNetCoreAnalyzers/test/WebApplicationBuilder/DisallowConfigureHostBuilderConfigureWebHostTest.cs:
```diff
@@ -0,0 +1,240 @@
+
+namespace Microsoft.AspNetCore.Analyzers.WebApplicationBuilder;
+
+public partial class DisallowConfigureHostBuilderConfigureWebHostTest
```
Sure, I'll take a quick look when I move the other bit of code into the shared source 👍

martincostello replied:
I had a quick look into this, but following the "quick start" docs, the test fails due to it not being able to find the dependencies that the `TestDiagnosticAnalyzerRunner` sets up. Seems a bit involved to get set up as part of this PR, so I'll leave that out of here.

Happy to help migrate it over later though once the initial pattern to get that stuff working is sorted out.

martincostello committed:
Implement analyzers for WebApplicationBuilder usage

Add analyzers that warn on incorrect usage of UseStartup(), Configure()
and ConfigureWebHost() when using WebApplicationBuilder.
Resolves #35814.

pranavkm commented:
/azp run
